### PR TITLE
DM-25450: Add Rowe statistics to SQuaSH

### PR DESCRIPTION
**Pull request by arunkannawadi**
This PR covers the calculation of several metrics related to Rho Statistics and their inclusion in the verify framework. There is an [associated PR](https://github.com/lsst/verify_metrics/pull/23) in `lsst/verify_metrics` that includes the definition of these new metrics and specifications.

### CODE REVIEW

File: python/lsst/pipe/analysis/analysis.py [hunk 1]
```diff
@@ -1099,10 +1099,12 @@ def plotQuiver(self, catalog, description, plotInfoDict, areaDict, log, cmap=plt
         yield Struct(fig=fig, description=description, stats=stats, statsHigh=None, dpi=150, style="quiver")
 
     def plotRhoStatistics(self, description, plotInfoDict, log, treecorrParams, stats,
-                          zpLabel=None, forcedStr=None, postFix="", flagsCat=None, uberCalLabel=None):
+                          zpLabel=None, forcedStr=None, postFix="", flagsCat=None, uberCalLabel=None,
+                          verifyJob=None):
```
Comment (MorganSchmitz) [inline python/lsst/pipe/analysis/analysis.py:1138]: Unrelated to rho_0 and very minor, but I think it would make more sense to add the `verifyJob` argument in 0aa731a14ca429645c85752f81e1750fe376f913 instead of here

**Review by MorganSchmitz — COMMENTED** [review]
Only looked at the $\rho_0$ commit but that looks good. Is there anywhere we actually explicitely define $\rho_0$ as the autocorrelation of size residual? If so, it would be good to link to that in one of the docstrings; if not, I think it's worth adding that there.

File: python/lsst/pipe/analysis/analysis.py (hunk 1 above)
Comment (arunkannawadi) [reply python/lsst/pipe/analysis/analysis.py:1138]: Agree. And I now define rho0 in the docstrings of RhoStatistics class.

File: python/lsst/pipe/analysis/utils.py [hunk 2]
```diff
@@ -525,9 +525,10 @@ class RhoStatistics(object):
 
     Returns
     -------
-    rhoStats : `dict` [`int', 'treecorr.GGCorrelation`]
-        A dictionary with keys 1..5, containing treecorr.GGCorrelation objects
-        corresponding to Rho statistic index
+    rhoStats : `dict [`int', 'treecorr.GGCorrelation`]
```
Comment (laurenam) [inline python/lsst/pipe/analysis/utils.py]: Something funky going on with the backticking here...

Also, no space below the dashed underline of Parameters, space before colon `psfColumn : `str``

I'd keep the short class summary shorter by removing the reference but adding that in an extended summary just below.

File: python/lsst/pipe/analysis/utils.py [hunk 3]
```diff
@@ -525,9 +525,10 @@ class RhoStatistics(object):
...
+        A dictionary with keys 0..5, containing one treecorr.KKCorrelation object (key 0)
```
Comment (laurenam) [inline python/lsst/pipe/analysis/utils.py]: I think you've exceeded the comment line length here (https://developer.lsst.io/python/style.html#docstring-and-comment-line-length-must-be-less-than-or-equal-to-79-columns).  Also, single backticks are still required in the description (https://developer.lsst.io/python/numpydoc.html#marking-up-parameter-names).

File: python/lsst/pipe/analysis/utils.py [hunk 4]
```diff
@@ -2185,10 +2188,55 @@ def calcQuartileClippedStats(dataArray, nSigmaToClip=3.0):
...
-    This is used to compute Rho statistics, given the appropriate shear-like
+    This is used to compute Rho0 statistics, given the appropriate spin-0 (scalar)
+    fields, usually fractional size residuals.
+
+    Parameters
+    ----------
+    ra : `numpy.array`
+       The right ascension values of entries in the catalog
+    dec : `numpy.array`
+       The declination values of entres in the catalog
```
Comment (laurenam) [inline python/lsst/pipe/analysis/utils.py]: entr\*i\*es

File: python/lsst/pipe/analysis/utils.py [hunk 5]
```diff
@@ -2185,10 +2188,55 @@ def calcQuartileClippedStats(dataArray, nSigmaToClip=3.0):
...
+    k2 : `numpy.array`, optional
+       The secondary scalar field.
+       Autocorrelation of the primary field is computed if None
+       (None by default)
+    raUnits : `str`, optional
+       Unit of the right ascension values (degrees by default)
+    decUnits : `str`, optional
+       Unit of the declination values (degrees by default)
+
+    Additionally, keyword arguments to treecorr.KKCorrelation may be passed
```
Comment (laurenam) [inline python/lsst/pipe/analysis/utils.py]: List this as **kwargs (as you do in `class RhoStatistics(object)` above).

File: python/lsst/pipe/analysis/utils.py [hunk 6]
```diff
@@ -2185,10 +2188,55 @@ def calcQuartileClippedStats(dataArray, nSigmaToClip=3.0):
...
+
+    Returns
+    -------
+    xy : `treecorr.KKCorrelation`
+       A treecorr.KKCorrelation object containing the correlation function
+    """
```
Comment (laurenam) [inline python/lsst/pipe/analysis/utils.py:2303]: Here and throughout, all descriptions above need backticks for referenced parameters and should end with a period.  It would be nice to include valid values for `ra/decUnits` (and a check on them to provide a useful error message, unless `treecore` can be trusted to provide that).

File: python/lsst/pipe/analysis/utils.py [hunk 7]
```diff
@@ -2185,10 +2188,55 @@ def calcQuartileClippedStats(dataArray, nSigmaToClip=3.0):
...
+        catB = treecorr.Catalog(ra=ra, dec=dec, k=k2, ra_units=raUnits,
+                                dec_units=decUnits)
+        # Calculate the cross-correlation
+        xy.process(catA, catB)
+
+    return xy
+
+def corrSpin2(ra, dec, g1a, g2a, g1b=None, g2b=None, raUnits="degrees", decUnits="degrees", **treecorrKwargs):
+    """ Function to compute correlations between atmost two spin-2 (shear-like) fields.
+
```
Comment (laurenam) [inline python/lsst/pipe/analysis/utils.py:2322]: No space after """ and I don't think "atmost" is a word 😉  (and check line length).

File: python/lsst/pipe/analysis/utils.py [hunk 8]
```diff
@@ -2235,3 +2284,29 @@ def corr2(ra, dec, g1a, g2a, g1b=None, g2b=None, raUnits="degrees", decUnits="de
         xy.process(catA, catB)
 
     return xy
+
+def measureRhoMetrics(rhoStat, D, operator=operator.le):
+    """ Convert Rho Statistics into a scalar metric.
+
```
Comment (laurenam) [inline python/lsst/pipe/analysis/utils.py:2373]: No space after """

File: python/lsst/pipe/analysis/utils.py [hunk 9]
```diff
@@ -2235,3 +2284,29 @@ def corr2(ra, dec, g1a, g2a, g1b=None, g2b=None, raUnits="degrees", decUnits="de
...
+    Parameters
+    ----------
+    rhoStat: `treecorr.GGCorrelation`
+        A correlation function corresponding to a Rho statistic
+    D: `float`
```
Comment (laurenam) [inline python/lsst/pipe/analysis/utils.py]: Can you use a more descriptive variable name here?

File: python/lsst/pipe/analysis/utils.py [hunk 10]
```diff
@@ -2235,3 +2284,29 @@ def corr2(ra, dec, g1a, g2a, g1b=None, g2b=None, raUnits="degrees", decUnits="de
...
+        A radial size
+    operator: An operator from the `operator` module, optional
```
Comment (laurenam) [inline python/lsst/pipe/analysis/utils.py]: This line should just include types (description goes in line below).

File: python/lsst/pipe/analysis/coaddAnalysis.py [hunk 11]
```diff
@@ -1701,7 +1713,6 @@ def runDataRef(self, patchRefList1, patchRefList2, subdir=""):
         if self.config.doPlotStarGalaxy:
             plotList.append(self.plotStarGal(forced, plotInfoDict, areaDict1, forcedStr=forcedStr,
                                              **plotKwargs1))
-
```
Comment (laurenam) [inline python/lsst/pipe/analysis/coaddAnalysis.py]: Unrelated line removal...

File: python/lsst/pipe/analysis/analysis.py [hunk 12]
```diff
@@ -1161,6 +1165,29 @@ def plotRhoStatistics(self, description, plotInfoDict, log, treecorrParams, stat
             rhoStats = rhoStatsFunc(good_catalog)
             plotRhoStats(axes, rhoStats)
 
+            if verifyJob is not None:
+                verifyJob.meta.update({"shapeAlgorithm": "HSM"})
+                measExtrasDictList = [{"name": kk, "value": vv, "label": None,
+                                       "description": "keyword arguments relevant to treecorr"}
+                                      for kk, vv in treecorrParams.items()]
+                measExtrasDictList.append(None)
+                catalogName = 'calibPsfUsed' if self.calibUsedOnly else 'allStars'
```
Comment (laurenam) [inline python/lsst/pipe/analysis/analysis.py]: May as well be consistent and use double quotes for strings here.

File: python/lsst/pipe/analysis/analysis.py [hunk 13]
```diff
@@ -1161,6 +1165,29 @@ def plotRhoStatistics(self, description, plotInfoDict, log, treecorrParams, stat
             rhoStats = rhoStatsFunc(good_catalog)
             plotRhoStats(axes, rhoStats)
 
+            if verifyJob is not None:
+                verifyJob.meta.update({"shapeAlgorithm": "HSM"})
```
Comment (laurenam) [inline python/lsst/pipe/analysis/analysis.py]: Since you are adding this to the metadata, I think you should leave the "HSM" out of the metric name so that it is not misleading if you were to compute the same metric but using a different shape algorithm.

Also, for consistency, you could use the `updateVerifyJob(job, metaDict=None, specsList=None)` utility function in `utils.py` to do the updating here.

File: python/lsst/pipe/analysis/coaddAnalysis.py [hunk 14]
```diff
@@ -500,9 +500,17 @@ def runDataRef(self, patchRefList, subdir="", cosmos=None):
                 else:
                     self.log.warn("Could not create match catalog for {:}.  Is "
                                   "lsst.meas.extensions.astrometryNet setup?".format(cat))
-
         self.allStats, self.allStatsHigh = savePlots(plotList, "plotCoadd", repoInfo.dataId, repoInfo.butler)
 
+        metaDict = {}
+        if repoInfo.camera:
```
Comment (laurenam) [inline python/lsst/pipe/analysis/coaddAnalysis.py]: Please add more relevant metadata such as the tract and filter.

File: python/lsst/pipe/analysis/coaddAnalysis.py [hunk 15]
```diff
@@ -500,9 +500,17 @@ def runDataRef(self, patchRefList, subdir="", cosmos=None):
                 else:
                     self.log.warn("Could not create match catalog for {:}.  Is "
                                   "lsst.meas.extensions.astrometryNet setup?".format(cat))
-
```
Comment (laurenam) [inline python/lsst/pipe/analysis/coaddAnalysis.py:554]: Unrelated line deletion?

File: python/lsst/pipe/analysis/utils.py [hunk 16]
```diff
@@ -2235,3 +2284,29 @@ def corr2(ra, dec, g1a, g2a, g1b=None, g2b=None, raUnits="degrees", decUnits="de
...
+        Typical values are gt, lt, ge, le. (le by default)
+
+    Returns
+    -------
+    avgRho: `float`
+        The value to track.
+    """
```
Comment (laurenam) [inline python/lsst/pipe/analysis/utils.py:2397]: Spaces before colons.

File: python/lsst/pipe/analysis/coaddAnalysis.py [hunk 17]
```diff
@@ -1390,25 +1398,29 @@ def plotRhoStatistics(self, catalog, plotInfoDict, zpLabel=None,
...
+                                      ("        Rho Statistics: "),
                                       shortName, self.config.analysis, flags=[], labeller=None
                                       ).plotRhoStatistics(shortName, plotInfoDict, self.log,
                                                           treecorrParams=self.config.treecorrParams,
                                                           stats=stats, zpLabel=zpLabel, forcedStr=forcedStr,
-                                                          uberCalLabel=uberCalLabel)
+                                                          uberCalLabel=uberCalLabel, verifyJob=self.verifyJob)
+
+        verifyJobFilename = repoInfo.butler.get("coaddAnalysis_verify_job_filename",
+                                                dataId=repoInfo.dataId)[0]
```
Comment (laurenam) [inline python/lsst/pipe/analysis/coaddAnalysis.py]: How is this working without an entry in `obs_subaru`'s policy file?

File: python/lsst/pipe/analysis/utils.py (hunk 6 above)
Comment (arunkannawadi) [reply python/lsst/pipe/analysis/utils.py:2303]: Value checking is done by `treecorr` and is actually less strict than it appears to be (e.g., 'deg', 'degrees' etc. are all acceptable values by `treecorr`)

File: python/lsst/pipe/analysis/analysis.py (hunk 13 above)
Comment (arunkannawadi) [reply python/lsst/pipe/analysis/analysis.py]: As mentioned [here](https://github.com/lsst/verify_metrics/pull/23#discussion_r480241922), the HSM name should probably stay.

File: python/lsst/pipe/analysis/coaddAnalysis.py (hunk 14 above)
Comment (arunkannawadi) [reply python/lsst/pipe/analysis/coaddAnalysis.py]: Using `plotInfoDict` which has all relevant information

File: python/lsst/pipe/analysis/coaddAnalysis.py [hunk 18]
```diff
@@ -516,6 +517,13 @@ def runDataRef(self, patchRefList, subdir="", cosmos=None):
         self.allStats, self.allStatsHigh = savePlots(plotList, "plotCoadd", repoInfo.dataId,
                                                      repoInfo.butler, subdir=subdir)
 
+        metaDict = {kk: plotInfoDict[kk] for kk in ("cameraName","filter","tract","visit","patch","rerun")}
```
Comment (laurenam) [inline python/lsst/pipe/analysis/coaddAnalysis.py]: Spaces after commas in list.

Since this is a per-tract metric, it does not make sense to include "patch" or "visit" in the metadata (I suppose one could imagine adding a "patches" or "patchList" entry that lists all the patches in the dataset, but as is, you'd only get a single one listed).  Also, the camera name has been referred to as "camera" (not "cameraName") in `colorAnalysis.py`.  If you feel strongly about "cameraName", please also change it in `colorAnalysis.py` (and, if you do go that route, you could also add the "rerun" entry there 😉 )

File: python/lsst/pipe/analysis/utils.py (hunk 6 above)
Comment (laurenam) [reply python/lsst/pipe/analysis/utils.py:2303]: Sounds good, but please use double ticks for strings throughout to be consistent.

File: python/lsst/pipe/analysis/analysis.py [hunk 19]
```diff
@@ -1096,10 +1096,12 @@ def plotQuiver(self, catalog, description, plotInfoDict, areaDict, log, cmap=plt
         yield Struct(fig=fig, description=description, stats=stats, statsHigh=None, dpi=150, style="quiver")
 
     def plotRhoStatistics(self, description, plotInfoDict, log, treecorrParams, stats,
-                          zpLabel=None, forcedStr=None, postFix="", flagsCat=None, uberCalLabel=None):
+                          zpLabel=None, forcedStr=None, postFix="", flagsCat=None, uberCalLabel=None,
+                          verifyJob=None):
         """ Plot Rho Statistics """
```
Comment (laurenam) [inline python/lsst/pipe/analysis/analysis.py]: Should be (with appropriate level of indentation):
"""Plot Rho Statistics.
"""

File: python/lsst/pipe/analysis/coaddAnalysis.py [hunk 20]
```diff
@@ -41,6 +41,7 @@
 import lsst.geom as geom
 import lsst.verify as verify
 
+
```
Comment (laurenam) [inline python/lsst/pipe/analysis/coaddAnalysis.py]: Extra blank line?

File: python/lsst/pipe/analysis/utils.py [hunk 21]
```diff
@@ -510,25 +511,29 @@ def __call__(self, catalog):
...
 
     Returns
     -------
-    rhoStats : `dict` [`int', 'treecorr.GGCorrelation`]
-        A dictionary with keys 1..5, containing treecorr.GGCorrelation objects
-        corresponding to Rho statistic index
+    rhoStats : `dict` [`int`, `treecorr.KKCorrelation` or
+                              `treecorr.GGCorrelation`]
+        A dictionary with keys 0..5, containing one treecorr.KKCorrelation
+        object (key 0) and five treecorr.GGCorrelation objects corresponding
```
Comment (laurenam) [inline python/lsst/pipe/analysis/utils.py]: Backticks: https://developer.lsst.io/python/numpydoc.html?highlight=backticks#marking-up-parameter-names

File: python/lsst/pipe/analysis/coaddAnalysis.py (hunk 20 above)
Comment (arunkannawadi) [inline python/lsst/pipe/analysis/coaddAnalysis.py]: I have no idea how this sneaked in here!

File: python/lsst/pipe/analysis/coaddAnalysis.py (hunk 18 above)
Comment (arunkannawadi) [reply python/lsst/pipe/analysis/coaddAnalysis.py]: Well, this can also be called in `visitAnalysis` and it runs through this implementation. I include it in the `metaDict` only if it is not None, so that only relevant info is passed whether it is called from `visitAnalysis` or `coaddAnalysis`. And I modified it to include `patchList`.

Comment (arunkannawadi) [reply python/lsst/pipe/analysis/coaddAnalysis.py]: The `plotInfoDict` has a `camera` object and its name as `cameraName`. It looks like the `camera` object is not serializable, nor do I want it in the metadata. To avoid confusion, I preferred to keep it as `cameraName`. I'll change it `colorAnalysis` in another ticket.

Comment (laurenam) [reply python/lsst/pipe/analysis/coaddAnalysis.py]: You mention that `visitAnalysis` can call this and that's why you left in the irrelevant `visit` metadata, but that's not the case. `visitAnalysis` does it's own `metaDict` setting here:
https://github.com/lsst-dm/pipe_analysis/blob/master/python/lsst/pipe/analysis/visitAnalysis.py#L527-L530
(and I don't know that you can guarantee that `visit` will resolve to `None` whenever you think it should...) 
 And do note the `camera` vs. `cameraName` entry.  If you INSIST on naming the tag `cameraName`, that will be anti-pattern to what is already in SQuaSH and will leave things inconsistent within `pipe_analysis`.  I would not want to live in this state while awaiting another ticket...I'd prefer keeping things consistent and, if you feel we *must* make the switch to `cameraName`, then that should happen everywhere on this ticket or atomically on a separate ticket.

Also, I don't think the `patchList` is really appropriate as a tag (how would a list of ~81 patches look in the dropdown menu...and is a `list` there even allowed?)  I think this might be more appropriate in the `measExtrasDictList`.

File: python/lsst/pipe/analysis/utils.py (hunk 6 above)
Comment (arunkannawadi) [reply python/lsst/pipe/analysis/utils.py:2303]: You mean double quotes for strings? I believe double ticks is for internal parameters.

File: python/lsst/pipe/analysis/coaddAnalysis.py (hunk 18 above)
Comment (arunkannawadi) [reply python/lsst/pipe/analysis/coaddAnalysis.py]: I wouldn't insist on naming it `cameraName`, but would have preferred it. I've changed it to call `camera` now. I also exclude any `patch` information, since it probably is run on all 81 patches by default.

**Comment by arunkannawadi** [comment]
I think I've addressed all the comments above. Apart from naming the metrics, if there are no more issues, I'll run the Travis.

File: python/lsst/pipe/analysis/plotUtils.py [hunk 22]
```diff
@@ -955,22 +955,27 @@ def plotRhoStats(axes, rhoStats):
     Parameters
     ----------
     axes : `list`
-        A list containing two `matplotlib.figure.ax` handles.
+        A list containing three `matplotlib.figure.ax` handles
```
Comment (laurenam) [inline python/lsst/pipe/analysis/plotUtils.py]: You dropped the period.

File: python/lsst/pipe/analysis/coaddAnalysis.py [hunk 23]
```diff
@@ -554,6 +554,15 @@ def runDataRef(self, patchRefList, subdir="", cosmos=None):
 
         self.allStats, self.allStatsHigh = savePlots(plotList, "plotCoadd", repoInfo.dataId,
                                                      repoInfo.butler, subdir=subdir)
+        metaDict = {kk: plotInfoDict[kk] for kk in ("cameraName", "filter", "tract", "patchList",
+                                                    "visit", "rerun")
+                    if plotInfoDict[kk] is not None}
+        self.verifyJob = updateVerifyJob(self.verifyJob, metaDict=metaDict, specsList=None)
```
Comment (laurenam) [inline python/lsst/pipe/analysis/coaddAnalysis.py:561]: Looks like you lost the changes here...

File: python/lsst/pipe/analysis/utils.py [hunk 24]
```diff
@@ -593,8 +599,9 @@ def __call__(self, catalog):
         e2SizeRes = e2*SizeRes
 
         # Package the arguments to capture auto-/cross-correlations for the
-        # Rho statistics.
-        args = {1: (e1Res, e2Res, None, None),
+        # Rho statistics
```
Comment (laurenam) [inline python/lsst/pipe/analysis/utils.py]: Convention is to end multiline comments with a period.

File: python/lsst/pipe/analysis/plotUtils.py [hunk 25]
```diff
@@ -955,22 +955,27 @@ def plotRhoStats(axes, rhoStats):
...
     rhoStats : `dict`
         A Python dictionary object with keys 1..5, each containing a
-        `treecorr.GGCorrelation object`.
+        `treecorr.GGCorrelation` object and key 0 containing a
+        `treecorr.KKCorrelation` object.
+
+    Returns
+    -------
+        `None`
```
Comment (laurenam) [inline python/lsst/pipe/analysis/plotUtils.py]: Return must be named (see description and warning at: https://developer.lsst.io/python/numpydoc.html#naming-return-variables).  Also, this function doesn't have an explicit return.

File: python/lsst/pipe/analysis/utils.py [hunk 26]
```diff
@@ -2309,3 +2366,44 @@ def corr2(ra, dec, g1a, g2a, g1b=None, g2b=None, raUnits="degrees", decUnits="de
...
+    rhoStat : `treecorr.GGCorrelation` or `treecorr.KKCorrelation` object
+        A correlation function corresponding to a Rho statistic.
+    thetaExtremum : `float`, optional
+        Extremum value of angular scale (in units of arcmin) to consider
+        for averaging ``rhoStat``.
+    operatorStr : `str`, optional
+        An optional comparison operator to specify angular scales smaller than
+        or larger than ``thetaExtremum``.
+        Allowed values are "<=", ">=", "<" or ">".
+
```
Comment (laurenam) [inline python/lsst/pipe/analysis/utils.py:2385]: Document the `Raises` (https://developer.lsst.io/python/numpydoc.html#py-docstring-raises).

File: python/lsst/pipe/analysis/analysis.py [hunk 27]
```diff
@@ -1173,7 +1177,9 @@ def plotRhoStatistics(self, description, plotInfoDict, log, treecorrParams, stat
                          style="RhoStats")
 
         if "ext_shapeHSM_HsmSourceMoments_xx" in self.catalog.schema:
-            figAxes = [plt.subplots(), plt.subplots()]  # first plot for Rho 1, 3, 4 and second for Rho 2, 5
+            figAxes = [plt.subplots(), plt.subplots(), plt.subplots()]
+            # first plot for Rho 1, 3, 4, second for Rho 2, 5
+            # and third for Rho 0
```
Comment (laurenam) [inline python/lsst/pipe/analysis/analysis.py]: Convention is to end multiline comments with a period.

File: python/lsst/pipe/analysis/plotUtils.py (hunk 25 above)
Comment (arunkannawadi) [reply python/lsst/pipe/analysis/plotUtils.py]: Since the return is not explicit, could I just drop the `Returns` part from the docstring?

Comment (laurenam) [reply python/lsst/pipe/analysis/plotUtils.py]: Yes, I think that's fine.  (We may want to eventually do a repo-wide standardizing of "to always return" or "to sometimes return", but not in this ticket!).

**Review by laurenam — APPROVED** [review]
